### PR TITLE
Relocate lab from SampleTissue to SampleAdditionalInfo

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/AbstractSample.java [hunk 1]
```diff
@@ -764,9 +764,6 @@ public Sample build() {
           checkArgument(sampleAdditionalInfo.getTissueType() != null,
               "SampleAdditionalInfo.tissueType must be provided to create a Sample.");
           return SampleImpl.sampleAnalyte(this);
-        } else if (sampleTissue != null) {
-          checkArgument(sampleTissue.getLab() != null, "SampleTissue.lab must be provided to create a SampleTissue");
```
Comment (callunity) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/AbstractSample.java]: Shouldn't the checkArgument line be the only thing removed?

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/SampleAdditionalInfoImpl.java [hunk 2]
```diff
@@ -275,6 +280,16 @@ public void setPrepKit(KitDescriptor prepKit) {
...
+
+  @Override
+  public void setLab(Lab lab) {
+    this.lab = lab;
+  }
+
+  @Override
   public Long getHibernateKitDescriptorId() {
     return kitDescriptorId;
   }
```
Comment (callunity) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/SampleAdditionalInfoImpl.java]: toString method? (not a requirement)

File: miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/SampleTissueDto.java [hunk 3]
```diff
@@ -126,9 +117,9 @@ public void setLabUrl(String labUrl) {
   @Override
   public String toString() {
     return "SampleTissueDto [bioBankId=" + bioBankId + ", cellularity=" + cellularity + ", createdById=" + createdById + ", createdByUrl="
-        + createdByUrl + ", creationDate=" + creationDate + ", id=" + id + ", labId=" + labId + ", labUrl=" + labUrl
-        + ", instituteTissueName=" + instituteTissueName + ", lastUpdated=" + lastUpdated + ", updatedById=" + updatedById
-        + ", updatedByUrl=" + updatedByUrl + ", url=" + url + "]";
+        + createdByUrl + ", creationDate=" + creationDate + ", id=" + id + ", labUrl=" + labUrl + ", instituteTissueName="
+        + instituteTissueName + ", lastUpdated=" + lastUpdated + ", updatedById=" + updatedById + ", updatedByUrl=" + updatedByUrl
+        + ", url=" + url + "]";
   }
```
Comment (callunity) [inline miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/SampleTissueDto.java]: labUrl needs to be taken out as well, and added in to SampleAdditionalInfoDto

**Comment by apmasell** [comment]
Fixed. The labUrl isn't set anywhere, BTW.

**Comment by tonydebat** [comment]
Lab is marked as NOT NULL which made sense when it was part of SampleTissue, since it would occur only once in the hierarchy. Now that it's part of SampleAdditionalInfo it will be present in every sample and entering the same value every time will be tiresome. Suggest we make Lab not a required field.

**Comment by apmasell** [comment]
Okay. Heather was in favour of making it required.

**Comment by callunity** [comment]
Tony makes a good point. We could set some internal logic that it must be included if it's at the top of the hierarchy, but it doesn't make sense to have it saved all the way down the hierarchy.

**Comment by apmasell** [comment]
Changed!

**Comment by callunity** [comment]
:+1:

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleAdditionalInfoService.java [hunk 4]
```diff
@@ -151,6 +156,7 @@ public SampleAdditionalInfo to(SampleAdditionalInfoDto sampleAdditionalInfoDto)
     authorizationManager.throwIfUnauthenticated();
     checkArgument(sampleAdditionalInfoDto.getSampleClassId() != null,
         "A SampleAdditionalInfo.sampleClassId must be provided to construct SampleAdditionalInfo.");
+    checkArgument(sampleAdditionalInfoDto.getLabId() != null, "A SampleAdditionalInfo.labId must be provided to construct SampleTissue.");
```
Comment (tonydebat) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleAdditionalInfoService.java]: Lab can be null, so this check should be removed.

**Comment by tonydebat** [comment]
:+1:

**Pull request merged**